mm2src/mm2_main/src/rpc/lp_commands/lr_swap/lr_impl.rs

```diff
@@ -0,0 +1,357 @@
+        let calc_total_price = |src_amount: U256, lr_swap: &ClassicSwapData, order: &RpcOrderbookEntryV2| {
+            let src_amount = mm_number_from_u256(src_amount);
+            let order_price = MmNumber::from(order.price.rational.clone());
+            let dst_amount = MmNumber::from(lr_swap.dst_amount.to_string().as_str());
+            if let Some(order_amount) = dst_amount.checked_div(&order_price) {
+                let total_price = src_amount.checked_div(&order_amount);
+                log::debug!("select_best_swap order.coin={} lr_swap.dst_amount(wei)={} order_amount(to fill order, wei)={} total_price(with LR)={}", 
+                    order.coin, lr_swap.dst_amount, order_amount.to_decimal(), total_price.clone().unwrap_or(MmNumber::from(0)).to_decimal());
+                total_price
+            } else {
+                None
+            }
+        };
+
+        self.inner
+            .values()
+            // filter out orders for which we did not get LR swap quotes and were not able to estimate needed source amount
+            .filter_map(|lr_data| {
+                if lr_data.src_amount.is_some() && lr_data.lr_swap_data.is_some() {
+                    Some((
+                        lr_data.src_amount.unwrap(),
+                        lr_data.lr_swap_data.as_ref().unwrap().clone(),
+                        lr_data.order.clone(),
+                    ))
+                } else {
+                    None
+                }
+            })
+            // calculate total price and filter out orders for which we could not calculate the total price
+            .filter_map(|(src_amount, lr_swap_data, order)| {
+                calc_total_price(src_amount, &lr_swap_data, &order)
+                    .map(|total_price| (lr_swap_data, order, total_price))
+            })
+            .min_by(|(_, _, price_0), (_, _, price_1)| price_0.cmp(price_1))
+            .map(|(lr_swap_data, order, price)| (lr_swap_data, order, price))
+            .ok_or(MmError::new(ApiIntegrationRpcError::LrSwapNotFound))
+    }
+}
+
+/// Finds the best swap path to buy order 's best "UTXO" coins, including LR quotes to sell my token for the rel tokens from the orders
+/// base_amount is amount of UTXO coins user would like to buy
+pub async fn find_best_fill_ask_with_lr(
+    ctx: &MmArc,
+    user_token: Ticker,
+    orders: &[RpcOrderbookEntryV2],
+    base_amount: &MmNumber,
+) -> MmResult<(ClassicSwapData, RpcOrderbookEntryV2, MmNumber), ApiIntegrationRpcError> {
+    let mut lr_data_map = LrDataMap::new_with_src_token(user_token, orders);
+    let _ = lr_data_map.update_with_contracts(ctx).await;
+    let _ = lr_data_map.calc_destination_token_amounts(ctx, base_amount).await;
+    let _ = lr_data_map.query_destination_token_prices(ctx).await?;
+    let _ = lr_data_map.estimate_source_token_amounts();
+    let _ = lr_data_map.run_lr_quotes(ctx).await?;
```

Comment:
To be honest I dont get, why you decided to have explicit result discarding, instead of just error propagation.

I think it should be like this
```rust
lr_data_map.update_with_contracts(ctx).await?;
lr_data_map.calc_destination_token_amounts(ctx, base_amount).await?;
lr_data_map.query_destination_token_prices(ctx).await?;
lr_data_map.estimate_source_token_amounts()?; 
lr_data_map.run_lr_quotes(ctx).await?;
```

Reply:
fixed thanks